File: TAREA1/bucle_ij_vs_ji.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

const MAX = 1024

// Go usa slices 2D (slice de slices) o un slice plano con indexado manual.
// Usamos un slice plano para garantizar contigüidad en memoria (como C/C++).
type Matrix [MAX * MAX]float64

var (
	A Matrix
	x [MAX]float64
	y [MAX]float64
)

// Acceso a la matriz: A[i][j] → A[i*MAX + j]  (row-major, igual que C)
func matGet(m *Matrix, i, j int) float64 { return m[i*MAX+j] }
func matSet(m *Matrix, i, j int, v float64) { m[i*MAX+j] = v }

// Inicializa A y x con valores aleatorios
func initialize() {
	rng := rand.New(rand.NewSource(42))
	for i := 0; i < MAX; i++ {
		x[i] = rng.Float64()
		for j := 0; j < MAX; j++ {
			matSet(&A, i, j, rng.Float64())
		}
	}
}

// Acceso contiguo en memoria → CACHE FRIENDLY
func loopIJ() {
	for i := range y {
		y[i] = 0
	}
	for i := 0; i < MAX; i++ {
		for j := 0; j < MAX; j++ {
			y[i] += matGet(&A, i, j) * x[j]
		}
	}
}


// Acceso con stride=MAX → CACHE UNFRIENDLY
func loopJI() {
	for i := range y {
		y[i] = 0
	}
	for j := 0; j < MAX; j++ {
		for i := 0; i < MAX; i++ {
			y[i] += matGet(&A, i, j) * x[j]
		}
	}
}

// Ejecuta la función 'runs' veces y retorna el tiempo promedio en ms
func benchmark(fn func(), runs int) float64 {
	var total time.Duration
	for r := 0; r < runs; r++ {
		start := time.Now()
		fn()
		total += time.Since(start)
	}
	return float64(total.Milliseconds()) / float64(runs)
}

func main() {
	initialize()

	runs := 5
	fmt.Println("==========================================")
	fmt.Printf("  Benchmark Go   |  MAX = %d  |  float64\n", MAX)
	fmt.Println("==========================================\n")

	timeIJ := benchmark(loopIJ, runs)
	timeJI := benchmark(loopJI, runs)
	ratio  := timeJI / timeIJ

	fmt.Printf("Bucle ij (filas, cache-friendly)  : %.3f ms\n", timeIJ)
	fmt.Printf("Bucle ji (columnas, cache-unfriend): %.3f ms\n", timeJI)
	fmt.Printf("\nRatio ji/ij = %.2fx\n", ratio)
	fmt.Printf("  → El bucle ji es ~%.2f veces MÁS LENTO\n\n", ratio)

	fmt.Println("Explicación:")
	fmt.Println("  Go (como C++) almacena arrays en row-major order.")
	fmt.Printf("  El bucle ij accede a A[i*MAX+j] de forma contigua (stride-1).\n")
	fmt.Printf("  El bucle ji accede a A[i*MAX+j] con stride=%d → muchos cache misses.\n", MAX)
}

File: TAREA1/bucle_ij_vs_ji.cpp


#include <iostream>
#include <chrono>
#include <cstdlib>
#include <cstring>
#include <iomanip>

#define MAX 1024

double A[MAX][MAX];
double x[MAX];
double y[MAX];

// Inicializa la matriz A y el vector x con valores aleatorios
void initialize() {
    for (int i = 0; i < MAX; i++) {
        x[i] = (double)rand() / RAND_MAX;
        for (int j = 0; j < MAX; j++) {
            A[i][j] = (double)rand() / RAND_MAX;
        }
    }
}

// Acceso a A[i][j]: secuencial en memoria → CACHE FRIENDLY
void loop_ij() {
    memset(y, 0, sizeof(y));
    for (int i = 0; i < MAX; i++)
        for (int j = 0; j < MAX; j++)
            y[i] += A[i][j] * x[j];
}

// Acceso a A[i][j] variando i en el bucle interno → CACHE UNFRIENDLY
void loop_ji() {
    memset(y, 0, sizeof(y));
    for (int j = 0; j < MAX; j++)
        for (int i = 0; i < MAX; i++)
            y[i] += A[i][j] * x[j];
}

// Ejecuta un benchmark: corre la función N veces y retorna promedio en ms
double benchmark(void (*func)(), int runs = 5) {
    double total = 0.0;
    for (int r = 0; r < runs; r++) {
        auto start = std::chrono::high_resolution_clock::now();
        func();
        auto end   = std::chrono::high_resolution_clock::now();
        total += std::chrono::duration<double, std::milli>(end - start).count();
    }
    return total / runs;
}

int main() {
    srand(42);
    initialize();

    std::cout << "==========================================\n";
    std::cout << "  Benchmark C++  |  MAX = " << MAX << "  |  double\n";
    std::cout << "==========================================\n\n";

    double time_ij = benchmark(loop_ij);
    double time_ji = benchmark(loop_ji);
    double ratio   = time_ji / time_ij;

    std::cout << std::fixed << std::setprecision(3);
    std::cout << "Bucle ij (filas, cache-friendly)  : " << time_ij << " ms\n";
    std::cout << "Bucle ji (columnas, cache-unfriend): " << time_ji << " ms\n";
    std::cout << "\nRatio ji/ij = " << ratio << "x\n";
    std::cout << "  → El bucle ji es ~" << ratio << " veces MÁS LENTO\n\n";

    std::cout << "Explicación:\n";
    std::cout << "  C++ almacena matrices en row-major order.\n";
    std::cout << "  El bucle ij accede a A[i][j] de forma contigua (stride-1).\n";
    std::cout << "  El bucle ji accede a A[i][j] con stride=" << MAX
              << " → muchos cache misses.\n";

    return 0;
}
